File: backend-node/src/infrastructure/seedData.ts
import { Money } from '../domain/money.js';
import { Transaction } from '../domain/transaction.js';
import { calculateMerchantFee } from '../services/feeCalculator.js';

/**
 * Port of Infrastructure/SeedData.cs. Same six fixture amounts as the .NET
 * version, deliberately chosen to land on fee-rounding fractions:
 * 150000, 99999, 1250500, 45000, 777777, and one Failed at 250000.
 */
export function seedTransactions(now: Date): Transaction[] {
  const daysAgo = (days: number) => new Date(now.getTime() - days * 24 * 60 * 60 * 1000);

  const rows: Array<{
    id: string;
    merchantId: string;
    gross: number;
    maskedPan: string;
    createdDaysAgo: number;
    status: 'Settled' | 'Failed';
  }> = [
    { id: 'trx_1001', merchantId: 'merchant_a', gross: 150000, maskedPan: '4111-XXXX-XXXX-1111', createdDaysAgo: 3, status: 'Settled' },
    { id: 'trx_1002', merchantId: 'merchant_a', gross: 99999, maskedPan: '4111-XXXX-XXXX-2222', createdDaysAgo: 10, status: 'Settled' },
    { id: 'trx_1003', merchantId: 'merchant_b', gross: 1250500, maskedPan: '5500-XXXX-XXXX-3333', createdDaysAgo: 1, status: 'Settled' },
    { id: 'trx_1004', merchantId: 'merchant_b', gross: 45000, maskedPan: '5500-XXXX-XXXX-4444', createdDaysAgo: 45, status: 'Settled' },
    { id: 'trx_1005', merchantId: 'merchant_c', gross: 777777, maskedPan: '4111-XXXX-XXXX-5555', createdDaysAgo: 89, status: 'Settled' },
    { id: 'trx_1006', merchantId: 'merchant_c', gross: 250000, maskedPan: '5500-XXXX-XXXX-6666', createdDaysAgo: 2, status: 'Failed' }
  ];

  return rows.map((r) => {
    const gross = Money.idr(r.gross);
    return new Transaction({
      id: r.id,
      merchantId: r.merchantId,
      gross,
      merchantFee: calculateMerchantFee(gross),
      maskedPan: r.maskedPan,
      createdAt: daysAgo(r.createdDaysAgo),
      status: r.status
    });
  });
}

File: backend/src/PaymentLab.Api/Infrastructure/SeedData.cs
using PaymentLab.Api.Domain;
using PaymentLab.Api.Services;

namespace PaymentLab.Api.Infrastructure;

public static class SeedData
{
    /// <summary>
    /// Deterministic fixtures. Amounts are chosen so that 0.7% MDR lands on
    /// awkward fractions — that is where rounding bugs show up.
    /// </summary>
    public static IReadOnlyList<Transaction> Transactions(DateTimeOffset now)
    {
        var fees = new FeeCalculator();

        Transaction Make(
            string id,
            string merchantId,
            long grossUnits,
            int daysAgo,
            string maskedPan,
            TransactionStatus status = TransactionStatus.Settled)
        {
            var gross = Money.Idr(grossUnits);
            return new Transaction
            {
                Id = id,
                MerchantId = merchantId,
                Gross = gross,
                MerchantFee = fees.CalculateMerchantFee(gross),
                AuthorizedAt = now.AddDays(-daysAgo),
                MaskedPan = maskedPan,
                Status = status
            };
        }

        return new List<Transaction>
        {
            Make("trx_1001", "mch_warungkopi", 150_000, 2, "455612******1234"),
            Make("trx_1002", "mch_warungkopi", 99_999, 9, "455612******7788"),
            Make("trx_1003", "mch_tokobuku", 1_250_500, 31, "521234******4321"),
            Make("trx_1004", "mch_tokobuku", 45_000, 120, "521234******9090"),
            Make("trx_1005", "mch_apotekseha", 777_777, 14, "400012******5555"),
            Make("trx_1006", "mch_apotekseha", 250_000, 5, "400012******6666", TransactionStatus.Failed)
        };
    }
}
